google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/rstudio/RStudioCredentialTester.java

```diff
@@ -0,0 +1,259 @@
+      HttpResponse response = httpClient.send(get(url).withEmptyHeaders().build());
+      canAcceptByCustomFingerprint =
+          response.status().isSuccess()
+              && response.headers().get(SERVER_HEADER).isPresent()
+              && response.headers().get(SERVER_HEADER).get().equals(RSTUDIO_HEADER)
+              && response
+                  .bodyString()
+                  .map(RStudioCredentialTester::bodyContainsRStudioElements)
+                  .orElse(false);
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Unable to query '%s'.", url);
+      return false;
+    }
+    return canAcceptByCustomFingerprint;
+  }
+
+  private static boolean bodyContainsRStudioElements(String responseBody) {
+    Document doc = Jsoup.parse(responseBody);
+    String title = doc.title();
+    String p = doc.body().getElementsByTag("p").first().outerHtml().split("<p>")[1].split("</p>")[0];
+
+    if (title.contains(RSTUDIO_UNSUPPORTED_BROWSER_TITLE) && p.contains(RSTUDIO_UNSUPPORTED_BROWSER_P)) {
+      logger.atInfo().log("Found RStudio endpoint");
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public ImmutableList<TestCredential> testValidCredentials(
+      NetworkService networkService, List<TestCredential> credentials) {
+
+    return credentials.stream()
+        .filter(cred -> isRStudioAccessible(networkService, cred))
+        .collect(toImmutableList());
+  }
+
+  private boolean isRStudioAccessible(NetworkService networkService, TestCredential credential) {
+    var url = buildTargetUrl(networkService, "auth-public-key");
+    try {
+      logger.atInfo().log("Retrieving public key");
+      HttpResponse response = httpClient.send(get(url).withEmptyHeaders().build());
+      Optional<String> body = response.bodyString();
+      String exponent = body.get().split(":")[0];
+      String modulus = body.get().split(":")[1];
+
+      url = buildTargetUrl(networkService, "auth-do-sign-in");
+      logger.atInfo().log(
+          "url: %s, username: %s, password: %s",
+          url, credential.username(), credential.password().orElse(""));
+      response = sendRequestWithCredentials(url, credential, exponent, modulus);
+
+      if (response.headers().get("Set-Cookie").isPresent()
+          && response.headers().get("Set-Cookie").get().trim().startsWith("user-id")) {
+        return true;
+      } else {
+        return false;
+      }
+    } catch (IOException
+        | NoSuchProviderException
+        | NoSuchAlgorithmException
+        | BadPaddingException
+        | IllegalBlockSizeException
+        | InvalidKeyException
+        | NoSuchPaddingException
+        | InvalidKeySpecException e) {
+      logger.atWarning().withCause(e).log("Unable to query '%s'.", url);
+      return false;
+    }
+  }
+
+  // This function is the same used from the javascript code in order to base64 encode the hexadecimal ciphertext. 
+  private String hexToBase64(String hex) {
```

Comment:
I think base encoding is already implemented in `com.google.common.io.BaseEncoding`. Have a look at `/tsunami-security-scanner-plugins/community/detectors/bigip_cve_2022_1388`. You should be able to remove it and replace with `BaseEncoding.base64().encode` to make the code shorter.

Reply:
I've tried to replace it with the function that you have said, but it's not working. Apparently rstudio performs some conversion from hex before converting it to base64. The function used in the code, is the translation of the one used in the client side code by rstudio. Let me know if i need to find an alternative